### PR TITLE
Can now specify the required versions of formatting utilities

### PR DESCRIPTION
**Pull request by tristan0x**
This PR handles the part of #89 saying that a project needs to be able to specify the versions of cmake-format and clang-format it supports.

See README.md additions for explanation

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -55,19 +55,14 @@ endfunction(cpp_cc_disable_precommit)
...
+  if(NOT ${CODING_CONV_PREFIX}_ClangFormat_REQUIRED_VERSION)
+    set(${CODING_CONV_PREFIX}_ClangFormat_REQUIRED_VERSION 7)
+  endif()
+  find_package(ClangFormat ${${CODING_CONV_PREFIX}_ClangFormat_REQUIRED_VERSION} REQUIRED)
   if(NOT ClangFormat_FOUND)
     message(SEND_ERROR "Could not find Clang-Format")
     return()
   endif()
-  if(ClangFormat_VERSION_MAJOR GREATER 6 AND ClangFormat_VERSION_MAJOR LESS 14)
-    set(${CODING_CONV_PREFIX}_ClangFormat_CONFIG_FILE clang-format-${ClangFormat_VERSION_MAJOR})
```
Comment (olupton) [inline cpp/CMakeLists.txt:64]: I'm wondering what happens with the configuration file now? A quick `git grep` says we still read from `..._ClangFormat_CONFIG_FILE` but that we don't set it any more.

I think the desired behaviour is that we use the newest `clang-format-X` file in this repo that is not tagged for a newer version of the `clang-format` executable than we find.
e.g. if the repo contains `clang-format-11` and `clang-format-12` config files then:
- `clang-format-11` executable should use the `clang-format-11` config
- `clang-format-12` executable should use the `clang-format-12` config
- `clang-format-13` executable should use the `clang-format-12` config

Comment (tristan0x) [reply cpp/CMakeLists.txt:64]: The 2 last commits should fulfill your expectations @olupton  :)

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -53,20 +53,74 @@ function(cpp_cc_disable_precommit)
...
-  else()
+  set(${CODING_CONV_PREFIX}_ClangFormat_CONFIG_FILE clang-format-${ClangFormat_VERSION_MAJOR})
```
Comment (olupton) [inline cpp/CMakeLists.txt:115]:
```suggestion
```
I think this is overwritten by the next line?

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -53,20 +53,74 @@ function(cpp_cc_disable_precommit)
...
+  cpp_cc_find_closest_config_file(
+    "clang-format" ${ClangFormat_VERSION_MAJOR} ${ClangFormat_MIN_VERSION}
+    ${CODING_CONV_PREFIX}_ClangFormat_CONFIG_FILE)
+  if(NOT ${CODING_CONV_PREFIX}_ClangFormat_CONFIG_FILE)
```
Comment (olupton) [inline cpp/CMakeLists.txt:119]: TIL `NOTFOUND` is a boolean false in CMake

**Review by olupton — COMMENTED** [review]
Thanks! One tiny nitpick on the diff, and another one here: now we have this support, can we delete the `clang-format-X` files that are just verbatim copies?
```
MD5 (clang-format-10) = f26fdd1a555b722f10d7c28907cb65db
MD5 (clang-format-11) = f26fdd1a555b722f10d7c28907cb65db
MD5 (clang-format-12) = 8d241f6dff62b819e1912fb8763b1c23
MD5 (clang-format-13) = 8d241f6dff62b819e1912fb8763b1c23
MD5 (clang-format-7) = edaae1de44ce824291a0d1f858eefd3d
MD5 (clang-format-8) = edaae1de44ce824291a0d1f858eefd3d
MD5 (clang-format-9) = 1765a9f1492a620cfde49ebda4727c40
```
looks like `clang-format-8`, `clang-format-11` and `clang-format-13` are all redundant?

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (tristan0x) [reply cpp/CMakeLists.txt]: Indeed. Thanks

**Comment by tristan0x** [comment]
> Thanks! One tiny nitpick on the diff, and another one here: now we have this support, can we delete the `clang-format-X` files that are just verbatim copies?
> 
> ```
> MD5 (clang-format-10) = f26fdd1a555b722f10d7c28907cb65db
> MD5 (clang-format-11) = f26fdd1a555b722f10d7c28907cb65db
> MD5 (clang-format-12) = 8d241f6dff62b819e1912fb8763b1c23
> MD5 (clang-format-13) = 8d241f6dff62b819e1912fb8763b1c23
> MD5 (clang-format-7) = edaae1de44ce824291a0d1f858eefd3d
> MD5 (clang-format-8) = edaae1de44ce824291a0d1f858eefd3d
> MD5 (clang-format-9) = 1765a9f1492a620cfde49ebda4727c40
> ```
> 
> looks like `clang-format-8`, `clang-format-11` and `clang-format-13` are all redundant?

Done in deb60d2f !
Thx @olupton

**Review by olupton — APPROVED** [review]

**Pull request merged**